Adds PCI bus driver implementation

gbowne1 opened the pull request:
Adding PCI driver to the kernel 

This is part of the required work for the kernel to be able to enumerate/detect devices directly off the PCI bus

There will need to be work done once this gets wired up to the kernel kmain.c including actual testing of devices on hardware and in emulation QEMU

gbowne1 committed:
Update display.h

updated header for display driver display.c and display.h this will need to be finished wired up. Old display driver would have done nothing.

gbowne1 committed:
Update display.c

Added the 95% completely wired up display driver implementation file

gbowne1 committed:
Create pci.h

gbowne1 committed:
Create pci.c

shoshta73 committed:
add configure script for setting up cross compilation tools

shoshta73 committed:
generate .build.env as part of configure script

shoshta73 committed:
update readme

shoshta73 committed:
lessen indirection in the makefile

shoshta73 committed:
initial implementation of klibc

fix linker error about ctx_switch

vmttmv committed:
fix stage2.asm: disk reads wait for BSY

shoshta73 committed:
fix RWX perms warnings in link step

gbowne1 committed:
Implement CPUID support check and CPU info printing

Added functions to check CPUID support and print CPU details.

gbowne1 committed:
Enhance cpu.h with Intel model definitions and struct

Added Intel model definitions and CPU info structure.

vmttmv committed:
Fix non-aligned disk reads in bootloader

vmttmv committed:
Establish well-defined read buffers for bl, implement error printing

shoshta73 committed:
fix vga clear section in `on_error`

gbowne1 committed:
Fix extern declaration for disk_read_sector function

gbowne1 committed:
Fix k_memcmp return logic and add disk_read_sector

Refactor k_memcmp to return correct difference and add disk_read_sector function.

vmttmv committed:
Remove types.c/.h, use klibc headers, amend stdbool.h, reimplement cpuid()

vmttmv committed:
Fix PDE/PTE definitions, header cleanup

- Fixes PDE/PTE definitions in kernel/paging.h
- removes memset declaration from kernel/utils.h, uses klibc string.h as
needed

gbowne1 committed:
Update vga.h

Add vga_init(); function prototype

gbowne1 committed:
Update vga.c

Change my_strlen to use our klibc strlen in klibc/include/string.h

vmttmv committed:
Fix includes for string.h/string_utils.h

shoshta73 committed:
add compile_commands.json genereation

shoshta73 committed:
add clangd config file

shoshta73 committed:
added editorconfig file

shoshta73 committed:
add .clang-format

shoshta73 committed:
add rules for formatting consecutives

gbowne1 committed:
Update gui.c

Add base gui implementation

gbowne1 committed:
Update gui.h

Adds gui base header

gbowne1 committed:
Create hid.h

Adding base HID device support for early HID standards 1.0

gbowne1 committed:
Create hid.c

Add bass HID implementation

gbowne1 committed:
Create ata.h

Add base ATA PIO mode driver so that filesystems like fat16 fat32 work.

gbowne1 committed:
Create ata.c

gbowne1 committed:
Update keyboard.c

remove static

gbowne1 committed:
Update keyboard.h

Add a extern const for the scancode map

gbowne1 committed:
Update keyboard.h

fixing missing includes and definition

gbowne1 committed:
Update keyboard.c

Fix missing definitions so theres nothing that would break the build

gbowne1 committed:
Update mouse.c

Make mouse_data non static

gbowne1 committed:
Update mouse.h

Fixes static

gbowne1 committed:
Update keyboard.h

fixed typo in header include

gbowne1 committed:
Update hid.c

Fixing print_hex error

vmttmv reviewed (changes requested):
Build errors:

```
kernel/pci.c: In function 'outl':
kernel/pci.c:5:5: error: 'asm' undeclared (first use in this function)
    5 |     asm volatile ("outl %0, %1" : : "a"(val), "Nd"(port));
      |     ^~~
kernel/pci.c:5:5: note: each undeclared identifier is reported only once for each function it appears in
kernel/pci.c:5:8: error: expected ';' before 'volatile'
    5 |     asm volatile ("outl %0, %1" : : "a"(val), "Nd"(port));
      |        ^~~~~~~~~
      |        ;
kernel/pci.c: In function 'inl':
kernel/pci.c:10:5: error: 'asm' undeclared (first use in this function)
   10 |     asm volatile ("inl %1, %0" : "=a"(ret) : "Nd"(port));
      |     ^~~
kernel/pci.c:10:8: error: expected ';' before 'volatile'
   10 |     asm volatile ("inl %1, %0" : "=a"(ret) : "Nd"(port));
      |        ^~~~~~~~~
      |        ;
```

`asm` is a GNU extension, as we're using `-std=c11`, `__asm__` should be used

gbowne1 committed:
Update pci.c

Fixed the inline assembly macro asm to __asm__

gbowne1 committed:
Update ata.c

Updated ata.c to include fixed BSY status register

gbowne1 committed:
Update ata.h

Updated header to match ata.c

gbowne1 committed:
Update pci.c

Fixed volatile to a keyword

vmttmv reviewed (commented):
This could use `io.h` instead of redefining `outl/inl`

gbowne1 committed:
Update pci.c

Change this to use io.h to define the in/out assembly instead of doing directly

shoshta73 committed:
Merge branch 'main' into gbowne1-addpci

shoshta73 reviewed (approved):
Tested locally, lgtm
